File: zemiao/78.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        n = len(nums)
        map = [0]*n
        res = []
        
        def dfs(begin):
          if begin == n:
            res.append([nums[i] for i in range(n) if map[i]])
            return
          
          dfs(begin+1)
          map[begin] = 1
          dfs(begin+1)
          map[begin] = 0
        
        dfs(0)
        return res

File: tnzhou/78.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> cur_result;
        if (nums.size() == 0) 
            return result;
        vector<int> temp {};
        result.push_back(temp);
        fetchAllResults(nums, cur_result, result, 0);
        return result;
    }

    void fetchAllResults(vector<int>& nums, vector<int>& cur_res, vector<vector<int>>& final_res, int this_idx) {
        if (this_idx >= nums.size()) {
            return;
        }
        cur_res.push_back(nums[this_idx]);
        final_res.push_back(cur_res);
        fetchAllResults(nums, cur_res, final_res, this_idx + 1);
        cur_res.pop_back();
        fetchAllResults(nums, cur_res, final_res, this_idx + 1);
    }

};
